ComponentKit/Core/CKComponentMemoizer.h

```diff
@@ -0,0 +1,148 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#import <Foundation/Foundation.h>
+
+#import <ComponentKit/CKComponentContext.h>
+#import <ComponentKit/CKComponentLayout.h>
+#import <ComponentKit/CKSizeRange.h>
+#import <ComponentKit/CKComponentSize.h>
+#import <ComponentKit/CKEqualityHashHelpers.h>
+
+#include <functional>
+
+// Aspect-oriented type erasure of hash/equals on a tuple
+struct CKMemoizationKey;
+
+id CKMemoize(CKMemoizationKey memoizationKey, id (^block)(void));
```

Comment:
You would need to cast the return type in that case; id allows you to say `MyComponent *a = CKMemoize()`...

Once we're building with Xcode 7, there are new features that allow the distinction of "allow assignment to any subclass of X".